shrs_line/src/buffer_history.rs

```diff
@@ -0,0 +1,89 @@
+// use std::fs::File;
+// use std::io::Write;
+
+use shrs_utils::cursor_buffer::{CursorBuffer, Location};
+
+pub trait BufferHistory {
+    fn next(&mut self, cb: &mut CursorBuffer);
```

Comment:
BufferHistory requires cursor pos and str, which are its only fields. We could pass them in as separate parameters. Buffer mutation would then have to happen in line.rs after checking if the operation was succesful or not.

Reply:
Another idea is to make DefaultBufferHistory internal to CursorBuffer, where buffer_history is a member of cursor_buffer. undo redo will call a method on cursor buffer which will internally use buffer history to set its own values.

Reply:
i was thinking more of implementing it as an [extension trait](http://xion.io/post/code/rust-extension-traits.html) where cursor buffer has a new function that modifies it's contents based on history, similalary to how `ViCursorBuffer` was implemented (this might be over-abstration though). The code is fine the way it is for now .